flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp

```diff
@@ -786,13 +797,59 @@ llvm::LogicalResult BroadcastAssignBufferization::matchAndRewrite(
   mlir::Value shape = hlfir::genShape(loc, builder, lhs);
   llvm::SmallVector<mlir::Value> extents =
       hlfir::getIndexExtents(loc, builder, shape);
-  hlfir::LoopNest loopNest =
-      hlfir::genLoopNest(loc, builder, extents, /*isUnordered=*/true,
-                         flangomp::shouldUseWorkshareLowering(assign));
-  builder.setInsertionPointToStart(loopNest.body);
-  auto arrayElement =
-      hlfir::getElementAt(loc, builder, lhs, loopNest.oneBasedIndices);
-  builder.create<hlfir::AssignOp>(loc, rhs, arrayElement);
+
+  bool isArrayRef =
+      mlir::isa<fir::SequenceType>(fir::unwrapRefType(lhs.getType()));
+  if (lhs.isSimplyContiguous() && extents.size() > 1 &&
+      (isArrayRef || isAllocatableArray(lhs.getType()))) {
+    // Flatten the array to use a single assign loop, that can be better
+    // optimized.
+    mlir::Value n = extents[0];
+    for (size_t i = 1; i < extents.size(); ++i)
+      n = builder.create<mlir::arith::MulIOp>(loc, n, extents[i]);
+    llvm::SmallVector<mlir::Value> flatExtents = {n};
+
+    mlir::Type flatArrayType;
+    mlir::Value flatArray = lhs.getBase();
+    if (isArrayRef) {
+      // Array references must have fixed shape, when used in assignments.
+      int64_t flatExtent = 1;
+      for (const mlir::Value &extent : extents) {
+        mlir::Operation *op = extent.getDefiningOp();
+        assert(op && "no defining operation for constant array extent");
+        flatExtent *= fir::toInt(mlir::cast<mlir::arith::ConstantOp>(*op));
+      }
+
+      flatArrayType =
+          fir::ReferenceType::get(fir::SequenceType::get({flatExtent}, eleTy));
+      flatArray = builder.createConvert(loc, flatArrayType, flatArray);
+    } else {
+      shape = builder.genShape(loc, flatExtents);
+      flatArrayType = fir::BoxType::get(
+          fir::HeapType::get(fir::SequenceType::get(eleTy, 1)));
+      flatArray = builder.create<fir::ReboxOp>(loc, flatArrayType, flatArray,
```

Comment:
I do not think you need this whole if-else construct.

`hlfir.designate` has the following builder:
```
void hlfir::DesignateOp::build(
    mlir::OpBuilder &builder, mlir::OperationState &result,
    mlir::Type result_type, mlir::Value memref, llvm::StringRef component,
    mlir::Value component_shape, llvm::ArrayRef<Subscript> subscripts,
    mlir::ValueRange substring, std::optional<bool> complex_part,
    mlir::Value shape, mlir::ValueRange typeparams,
    fir::FortranVariableFlagsAttr fortran_attrs)
```

I believe you may just pass the array's base address as memref casted to the flat-array type, and the 1D shape built from the `flatExtents`.

Reply:
I tried this, but it fails on MLIR verification. The error was something like:
`memref must be any HLFIR variable type.`

`hlfir::isFortranVariableType()` doesn't consider a reference to an array with dynamic shape to be a valid Fortran variable. That's why I added the code to cast ND fixed shapes to 1D fixed shapes and to rebox ND dynamic shapes.

Reply:
Okay, my bad.  You are right, it won't work.